### PR TITLE
Add basic link resolve spec

### PR DESCRIPTION
**Pull request by yufeih**
#2699 #2782

### CODE REVIEW

File: src/docfx/docset/Document.cs [hunk 1]
```diff
@@ -55,14 +55,19 @@ internal class Document
         /// </summary>
         public string OutputPath { get; }
 
-        public Document(Docset docset, string filePath)
+        /// <summary>
+        /// Intentionally left as private. Use <see cref="Document.TryCreate(Docset, string)"/> instead.
+        /// </summary>
+        internal Document(Docset docset, string filePath)
```
Comment (superyyrrzz) [inline src/docfx/docset/Document.cs]: >internal [](start = 8, length = 8)

Should be private according to comments?

File: docs/specs/build/resolve.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+outputs:
+  docs/dir-a/a.json: '{ "content": "<p>Link to <a href=\"../dir-b/b\">b</a></p>" }'
+  docs/dir-b/b.json: '{ "content": "<p>Link to <a href=\"../dir-a/a\">a</a></p>" }'
+  build.manifest:
+---
+# Path starting with ~ resolves from docset root directory
+inputs:
+  docfx.yml:
+  docs/dir-a/a.md: Link to [b](~/docs/dir-b/b.md)
+  docs/dir-b/b.md: Link to [a](~docs/dir-a/a.md)
```
Comment (OsmondJiang) [inline docs/specs/build/resolve.yml]: >~docs [](start = 31, length = 5)

do we really want to support that?

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 3]
```diff
@@ -132,14 +132,14 @@ private static void ResolveTocModelItems(List<TableOfContentsInputItem> tocModel
...
                         {
-                            tocModelItem.Items = LoadInputModelItems(referencedTocContent, isYamlToc, referenceTocFilePath, rootPath, resolveContent, resolveHref, parents);
+                            tocModelItem.Items = LoadInputModelItems(referencedTocContent, isYamlToc, referenceTocFilePath, resolveContent, resolveHref, parents);
                             tocModelItem.Href = topicHref;
                         }
                     }
                     else
                     {
                         tocModelItem.Href = string.IsNullOrEmpty(topicHref)
-                            ? resolveHref?.Invoke(filePath, href, rootPath) ?? tocModelItem.Href
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: >rootPath [](start = 66, length = 8)

where is rootpath?

File: docs/specs/build/resolve.yml (hunk 2 above)
Comment (yufeih) [reply docs/specs/build/resolve.yml]: This link is not resolved. we should give a warning. I left a TODO in code to add warnings later

---
In reply to: [193270047](https://github.com/dotnet/docfx/pull/2839#discussion_r193270047) [](ancestors = 193270047)

File: src/docfx/build/toc/parser/TableOfContentsParser.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: i think the interface should contains all the necessary inputs to generate expected output, current resolvehref can't give me enough information for that.

---
In reply to: [193270998](https://github.com/dotnet/docfx/pull/2839#discussion_r193270998) [](ancestors = 193270998)

File: src/docfx/resolve/Resolve.cs [hunk 4]
```diff
@@ -2,82 +2,88 @@
...
 
+            // Make result relative to `resultRelativeTo`
             resolvedHref = PathUtility.GetRelativePathToFile(resultRelativeTo.SiteUrl, file.SiteUrl).Replace('\\', '/');
 
-            return (Uri.EscapeUriString(resolvedHref) + fragment + query, file);
+            return (resolvedHref + fragmentQuery, file);
         }
 
-        private static Document TryResolveHref(this Document relativeTo, string href)
+        private static (Document file, string fragmentQuery) TryResolveFile(this Document relativeTo, string href)
```
Comment (superyyrrzz) [inline src/docfx/resolve/Resolve.cs]: >TryResolveFile [](start = 61, length = 14)

TryResolveHref? as the parameter is href

File: src/docfx/resolve/Resolve.cs [hunk 5]
```diff
@@ -2,82 +2,88 @@
...
         {
             if (string.IsNullOrEmpty(href))
             {
                 return default;
             }
 
-            if (!HrefUtility.IsRelativeHref(href))
+            var (path, fragment, query) = HrefUtility.SplitHref(href);
```
Comment (superyyrrzz) [inline src/docfx/resolve/Resolve.cs]: >path [](start = 17, length = 4)

need url decode href to get path?

File: src/docfx/resolve/Resolve.cs (hunk 4 above)
Comment (yufeih) [reply src/docfx/resolve/Resolve.cs]: that'll have conflict with the above TryResolveHref. The naming suggest the return value, so TryResolveHref returns an href, TryResolveFile returns a document, TryResolveContent returns the text.

---
In reply to: [193278780](https://github.com/dotnet/docfx/pull/2839#discussion_r193278780) [](ancestors = 193278780)

File: src/docfx/resolve/Resolve.cs (hunk 5 above)
Comment (yufeih) [reply src/docfx/resolve/Resolve.cs]: Right, I left a TODO in Document to handle URL encoding/decoding. They will be handled later.

---
In reply to: [193278818](https://github.com/dotnet/docfx/pull/2839#discussion_r193278818) [](ancestors = 193278818)

**Pull request merged**